### PR TITLE
net/devif: check the net device before use

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/devif: check the net device before use

Signed-off-by: chao an <anchao@xiaomi.com>

## Impact

N/A

## Testing

tcp/udp test

### CODE REVIEW

File: net/devif/devif_iobsend.c [hunk 1]
```diff
@@ -56,67 +57,53 @@ void devif_iob_send(FAR struct net_driver_s *dev, FAR struct iob_s *iob,
...
-  if (dev == NULL || len == 0 || len > limit)
-#else
   if (dev == NULL || len == 0)
-#endif
     {
-      if (dev->d_iob == NULL)
-        {
-          iob_free_chain(iob);
-        }
+      ret = -EINVAL;
```
Comment (acassis) [inline net/devif/devif_iobsend.c:64]: Case dev==NULL shouldn't it return -ENODEV instead of -EINVAL?

**Review by acassis — APPROVED** [review]

File: net/devif/devif_iobsend.c (hunk 1 above)
Comment (anchao) [reply net/devif/devif_iobsend.c:64]: Done

File: net/devif/devif_send.c [hunk 2]
```diff
@@ -68,47 +69,50 @@
...
-    }
-  else
+  ret = iob_trycopyin(dev->d_iob, buf, len, offset, false);
+  if (ret != len)
     {
-      dev->d_sndlen = 0;
+      netdev_iob_release(dev);
     }
 
-  if (dev->d_sndlen != len)
```
Comment (xiaoxiang781216) [inline net/devif/devif_send.c:109]: let's return here to avoid the check(ret < 0) after errout

File: net/devif/devif_iobsend.c [hunk 3]
```diff
@@ -56,67 +57,59 @@ void devif_iob_send(FAR struct net_driver_s *dev, FAR struct iob_s *iob,
...
 
+  dev->d_sndlen = len;
+
 #ifdef CONFIG_NET_TCP_WRBUFFER_DUMP
   /* Dump the outgoing device buffer */
 
   lib_dumpbuffer("devif_iob_send", dev->d_appdata, len);
 #endif
+
+errout:
```
Comment (xiaoxiang781216) [inline net/devif/devif_iobsend.c:108]: let's return here to avoid the check(ret < 0) after errout

File: net/devif/devif_send.c (hunk 2 above)
Comment (anchao) [reply net/devif/devif_send.c:109]: Done

File: net/devif/devif_iobsend.c (hunk 3 above)
Comment (anchao) [reply net/devif/devif_iobsend.c:110]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@anchao 

I noticed that the renew command with lm3s6965-ek:discover on QEMU-7.1 does not work with this PR.

```
nsh> uname -a
NuttX  10.4.0 7fce145b30 Jan 31 2023 20:35:26 arm lm3s6965-ek
nsh> ps
  PID GROUP PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK   STACK   USED  FILLED COMMAND
    0     0   0 FIFO     Kthread N-- Ready              00000000 001000 000472  47.2%  Idle Task
    1     1 224 RR       Kthread --- Waiting  Semaphore 00000000 001992 000276  13.8%  hpwork 0x20001cd0
    2     2  60 RR       Kthread --- Waiting  Semaphore 00000000 001992 000276  13.8%  lpwork 0x20001ce8
    3     3 100 RR       Task    --- Running            00000000 002000 001160  58.0%  nsh_main
    4     4 100 RR       Task    --- Waiting  Semaphore 00000000 002008 000564  28.0%  telnetd
nsh> ifconfig
eth0	Link encap:Ethernet HWaddr 00:e0:de:ad:be:ef at UP
	inet addr:10.0.0.2 DRaddr:10.0.0.1 Mask:255.255.255.0

             IPv4   TCP   UDP  ICMP
Received     0000  0000  0000  0000
Dropped      0000  0000  0000  0000
  IPv4        VHL: 0000   Frg: 0000
  Checksum   0000  0000  0000  ----
  TCP         ACK: 0000   SYN: 0000
              RST: 0000  0000
  Type       0000  ----  ----  0000
Sent         0000  0000  0000  0000
  Rexmit     ----  0000  ----  ----
nsh> renew eth0
ERROR: dhcpc_request() failed
```

**Comment by anchao** [comment]
> @anchao
> 
> I noticed that the renew command with lm3s6965-ek:discover on QEMU-7.1 does not work with this PR.

Could you please share the qemu command line?

**Comment by masayuki2009** [comment]
@anchao 

>Could you please share the qemu command line?

```
/home/ishikawa/opensource/QEMU/qemu-7.1/build/arm-softmmu/qemu-system-arm -net nic,model=stellaris -net user,hostfwd=tcp::20023-:23,hostfwd=tcp::20021-:21,hostfwd=tcp::25001-:5001 -M lm3s6965evb -kernel ./nuttx -nographic
```

**Comment by anchao** [comment]
> @anchao
> 
> > Could you please share the qemu command line?
> 
> ```
> /home/ishikawa/opensource/QEMU/qemu-7.1/build/arm-softmmu/qemu-system-arm -net nic,model=stellaris -net user,hostfwd=tcp::20023-:23,hostfwd=tcp::20021-:21,hostfwd=tcp::25001-:5001 -M lm3s6965evb -kernel ./nuttx -nographic
> ```

@masayuki2009  thank you, please review https://github.com/apache/nuttx/pull/8380

**Pull request merged**